### ISSUE TITLE
reg-name ABNF is unspecified

### ISSUE THREAD

**Issue by sleevi** [w3c-cg/web-nfc#350]
#278 introduced the `reg-name` within the ABNF, and prosaically describes it 
> a [registrable domain](https://url.spec.whatwg.org/#host-registrable-domain) owned by the issuing organization

However, the actual ABNF for `reg-name` is unspecified, and this can lead to some ambiguity. For example, the domain name system is 8-bit clean; `\x00.domain.example` is a valid domain name, which is made up of the labels [(1 byte, `\x00`), (6 bytes, `domain`), (7 bytes, `example`)] when transmitted on DNS wire format.

It's true that the host record type (A, or AAAA) uses "preferred name syntax" of letters, digits, and hyphens, but even that has caveats. For example, the URL Standard (and its predecessors) allowed underscores (`_`) within host names, unescaped. Many user agents would pass this on to their name resolution libraries (like `gethostbyname`), which would not enforce that input comes in preferred name syntax, and would wire-encode such names when looking up A and AAAA records.

Unfortunately, it's not yet uncommon to see underscores in hostnames, and [only recently](https://cabforum.org/2018/11/12/ballot-sc-12-sunset-of-underscores-in-dnsnames/) (and to some disruption [1](https://bugzilla.mozilla.org/show_bug.cgi?id=1516561), [2](https://bugzilla.mozilla.org/show_bug.cgi?id=1517617), [3](https://bugzilla.mozilla.org/show_bug.cgi?id=1515788), [4](https://bugzilla.mozilla.org/show_bug.cgi?id=1516453), [5](https://bugzilla.mozilla.org/show_bug.cgi?id=1516599)) have browsers reiterated that certificates for those domains should never exist.

Thus, it would be better to precisely describe the character space for `reg-name` in no ambiguous terms. If it's meant that `reg-name` should be imported from RFC 3986, that would be clearer. However, that ABNF and algorithm described is one known to cause significant pain for consistent, cross-browser implementation. ABNF-style approaches, and their failure to adequately describe processing models and error handling, are why things like the [URL Standard](https://url.spec.whatwg.org/) exist.

To better ensure interoperability, please consider writing out a clear processing model intended for handling these sorts of tags, and a clear processing model intended for interoperably encoding these tags. Of particular concern and consideration is the need to make sure that it's clear the rules for distinguishing the "domain" portion and any extended attributes.

As a concrete example, one implementation issue I could foresee is the use of this API on IPv6 addresses, which, in URL presentation form, are encoded such as `[::1]`. It's subtle, but if we assume `reg-name` is meant to refer to the domain name, then this feature should not be allowed to work on such hosts. However, if a user agent doesn't prohibit this/explicitly check, they might permit a tag of `[::1]:foo` to be created. Is that valid or not? When another user agent encounters that tag, if they merely scan for `:`, they may parse the tag as (`[`, `:1]:foo`)(if they look for the first `:`), or (`[::1]`, `foo`) if they look for the last `:`. However, since `:` is allowed in the `other` of the NFC tag, it means a tag like `[::1]:foo:bar` is even more confusing to parsers!

Hopefully, the precision of the URL Standard provides a good model to concretely provide interoperable algorithms and encoding rules, to avoid many of the issues I highlighted above.

**Comment by zolkis** [comment]
> To better ensure interoperability, please consider writing out a clear processing model intended for handling these sorts of tags, and a clear processing model intended for interoperably encoding these tags.

If we are talking about NFC interoperability (with any NFC device in general), we should rely on the NFC Forum specifications.

If we are talking about ensuring interoperability between web apps/pages that may use NFC as intermediate storage (for tags) or bitpipe (for push), then (provided the NFC specifications are respected) we are free to define relevant policies for encoding, URL handling and content handling.
Probably we could use some help in formulating these the right way and will likely deal with this after TPAC.

**Comment by sleevi** [comment]
So, I'm not sure where the line is between them, but I was specifically talking about the `NDEFRecordType` external type processing model (specifically, Section 2.4.5 "External Type Records (TNF 4)").

It's unclear if the language is being imported from the NFC Forum Technical specifications, or if it's language being explicitly described for interoperability with UAs. Section 7.4.4, "Store site URL as record identifier when writing data" made me believe that this was something specific to Web NFC, which is why I wanted to make sure there was a clearly defined processing model associated with this, since the current settings object may not have a registrable domain (it may be an IP).

The processing model in Section 7.4.3, "Permissions controls", suggest that there's matching based on ID fields, which means that there's a processing element to parsing those responses.

**Comment by kenchris** [comment]
NFC spec says to avoid using IP addresses whenever possible, but when used the literal format for IPv6 addresses in URLs (RFC 2732) should be supported. I am not sure what that means for us spec wise.

Also the spec states that it does not define any equivalence rules for URLs, but due to inconsistencies in parsers, it is recommended that generators (like Web NFC) only rely on the most rudimentary equivalence rules defined in RFC 2986

**Comment by kenchris** [comment]
Also spec says "The External Type Name MUST be formed by taking the domain name of the issuing
organization, adding a colon, and then adding the type name as managed by the organization."

Maybe we can avoid enforcing restrictions and instead add a note about this? @zolkis

**Comment by leonhsl** [comment]
About `An external type name` at https://w3c.github.io/web-nfc/#the-record-type-string,
1. it still says 

> The reg-name value is a registrable domain

2. it's confusing that the section for `An external type name` is talking about well-known types.

> And additional ABNF exists for well-known type records:
wkt-type             = (ALPHA / DIGIT) *(ALPHA / DIGIT / other)

**Comment by beaufortfrancois** [comment]
@zolkis Can you address latest comment from @leonhsl at https://github.com/w3c/web-nfc/issues/350#issuecomment-552276574?

**Comment by beaufortfrancois** [comment]
@zolkis (gentle ping)

**Comment by zolkis** [comment]
Since now we have a validation for external types in
https://w3c.github.io/web-nfc/#validating-external-type ,
We could move the ABNF's form the informative section above to the algorithm, or remove it (if the algorithm is enough). 
WKT doesn't really need an ABNF and I'd remove it.

**Comment by kenchris** [comment]
It might be useful for people reading the spec though but it could be non nominative

**Comment by beaufortfrancois** [comment]
What are the next steps?

**Comment by zolkis** [comment]
I will attempt a PR :)

**Comment by leonhsl** [comment]
Some problems in 9.11.1 Creating NDEF record https://w3c.github.io/web-nfc/#creating-ndef-record:
  - Step 4, several standardized record types are missing.
  - Step 5, the following description is wrong. The created NDEF message is supposed to be used as the current record's payload.

> If record's data is of type NDEFMessageInit, then return the result of running the create NDEF message given record's data.
  - Step 6 seems to be copied from Step 5, not sure whether it wants to address local type record or Sp record, anyway it needs to reconsider.

**Comment by leonhsl** [comment]
About https://w3c.github.io/web-nfc/#validating-external-type,

  - it does not seem to be validating the following rule for external types, so I want to confirm whether the following rule is already stale?
> An external type name
A DOMString representing a custom type for the external type record. The type must follow the external type name ABNF.
ext-type             = reg-name ":" custom-type
custom-type          = 1*(ALPHA / DIGIT / other)
DIGIT                = %x30-39
ALPHA                = %x41-5A / %x61-7A   ; A-Z / a-z
other                = "(" / ")" / "+" / "," / "-" / ":" / "=" /
                       "@" / ";" / "$" / "_" / "!" / "*" / "'" / "."

  - Step 2, do we need to make sure that `input` contains a colon `:` before returning it?

> If input does not start with colon (':'), return input.

Because of the definition

> The [NFC-RTD] specifies that external types MUST contain the domain name of the issuing organization, a colon and a type name that is at least one character long.

**Comment by zolkis** [comment]
See #474 for some related changes. It needs opinions @leonhsl  (I made a few comments on my own). Then, does that ABNF definition help you, or should we remove it?

> Some problems in 9.11.1 Creating NDEF record https://w3c.github.io/web-nfc/#creating-ndef-record:
Step 4, several standardized record types are missing.

The PR does not add missing type creations yet, that will come in another PR, but only for "smart-poster", since "absolute-url", "signature" and "handover" will not be included in the OT (maybe not even Sp).

> Step 5, the following description is wrong. The created NDEF message is supposed to be used as the current record's payload. 
> Step 6 seems to be copied from Step 5, not sure whether it wants to address local type record or Sp record, anyway it needs to reconsider.

Good catch, let's track this in a separate issue.

**Comment by zolkis** [comment]
@sleevi, the sections related to this issue have been rewritten.
I wonder could you please take another look and check if we can close this.

Now we have [validation steps](https://w3c.github.io/web-nfc/#validating-external-type) for external type, including its domain part.

**Comment by beaufortfrancois** [comment]
@sleevi (gentle ping)

**Comment by sleevi** [comment]
Thanks for adding, and thanks for @beaufortfrancois pinging, as this slipped past due to the holidays (and holiday deadlines :D)

Overall, I think the approach in having an explicit validation step is good, in that it makes it much clearer (and easier to write tests for interop issues).

> 1. If input is not a USVString or it is empty, or its length exceeds 255 bytes, return false.

Is 255 bytes a limit from NFC-RTD? Or is this trying to limit for domain names? If it's trying to limit for domain names, I believe you want the 'size check' to be after Step 5. Prior to that, you're dealing in USVString, which corresponds to Unicode scalar values, which means 'bytes' vs' characters' is messy, to say the least! It also means that IDNs that are < 255 bytes in U-Label form but > 255 bytes in A-Label form would be incorrectly permitted.

> 2. Let domain be the input from the start of input up to but excluding the last occurrence of U+003A (:), or null if that is not found.

I don't have access to the NFC-RTD specification, so I'm going based on how I read Step 7 (so hopefully this makes sense). My understanding is that `type` MUST NOT contain `'U+003A (:)`, that this comes from the NFC-RTD specification, and so there's no risk of a type being `domain.example:foo:bar` with the intended _domain_ of `domain.example` and the intended _type_ of `foo:bar`, right?

Apologies if this is something answered in NFC-RTD, I'm not used to Web specifications making normative dependencies on pay-to-access specs.

> Let asciiDomain be the result of running domain to ASCII on domain.

You've omitted the optional boolean beStrict, which means it defaults to false. This permits certain [implementation-dependent characters](https://unicode.org/reports/tr46/#UseSTD3ASCIIRules). It's unclear if that was intended here - that is, the relationship between the *external type* and the current origin are unclear me, but I would suspect that NFC-RTD makes a normative dependency on DNS, which means that beStrict should be true, even if UAs allow the current domain to contain a `U+005F (_) LOW LINE (underbar)`. My general advice is "be strict in what you accept" - if you're validating caller textual input or NFC tag input (versus the current origin/domain), then omitting `_` seems entirely reasonable.


The only other bit I would flag, and I fully admit I'm not an expert here and so you'll want to find someone better versed in the Web Platform idioms, is where/how an error should be signaled for an unsupported `recordType`. Right now, if I understand correctly, I can create a 'valid' `NDEFRecord`  via `NDEFRecordInit`, but whose `recordType` is invalid/unparsable. That error is not actually signaled back to the consumer until they try [`NDEFWriter.push`](https://w3c.github.io/web-nfc/#the-push-method), which invokes the [`create NDEF message`](https://w3c.github.io/web-nfc/#dfn-create-ndef-message) algorithm, which invokes [`create NDEF record`](https://w3c.github.io/web-nfc/#dfn-create-ndef-record), which invokes [`validate external type`](https://w3c.github.io/web-nfc/#dfn-validate-external-type).

Because you [expose the constructor](https://w3c.github.io/web-nfc/#idl-index) for `NDEFRecord`, should you be parsing/validating the type then? The URL class, for better or worse, has setters which do the parsing/validation. I'm not sure why the `NDEFRecord` type is read-only (maybe that's the current best practice), but it seems that you at least want to validate during that constructor as well, to ensure a client only encounters well-formed `NDEFRecord`s

**Comment by zolkis** [comment]
Thanks @sleevi for the helpful insights, almost each of which uncovers bugs in the spec.

> Is 255 bytes a limit from NFC-RTD? 

It comes from the NDEF spec. Type length field is 8 bits, which limits type field to 255 bytes (octets).
Any encoding must fit in this space. Underlying platforms should take care of truncating or throwing away input longer than this, so this is just a safety step here.

> My understanding is that type MUST NOT contain `'U+003A (:)`, that this comes from the NFC-RTD specification, and so there's no risk of a type being domain.example:foo:bar with the intended domain of domain.example and the intended type of foo:bar, right?

A good observation, and yeah, welcome to the hive :). I try to summarize the main things. The  NFC RTD says:

> Record type names SHALL be formed of characters from of the US ASCII character set.
Characters in the range [0-31] and 127 decimal SHALL NOT be used in record type names.
(...)
Record type names may be specified in several formats, called Type Name Formats, as signified
by the TNF field of the NDEF record header. 

Out of these, the NFC RTD defines the format for the NFC Forum standardized type names.
There is some confusion there. The RTD spec mandates type names be a URN ([RFC 2141], with the namespace identifier “nfc”, namespace specific string "wkt:". However, when encoded in an NDEF message, the Well Known Type MUST be written as a relative-URI construct [RFC 3986], omitting the NID and the “wkt:” prefix. 

The ABNF for RTD type names takes the _first_ occurrence of US-ASCII `:` as delimiter between `dns-part` and `name-part` of an external type name. The RTD spec allows `:` in the `name-part`. 
This would allow `domain.example:foo:bar` with `dns-part` being `domain.example` and `name-part` being `foo:bar`.

However, the NDEF spec says the literal format for IPv6 addresses in URIs as described by RFC
2732 SHOULD be supported.

Since we have a conflict here, we made the choice of disallowing `:` in the `name-part` and take the _last_ occurrence of `:` as delimiter, which IIRC was done as quick fix to a review comment. 

Probably what we should do instead is figure out if `dns-part` is literal IPv6 and take the first `:` as delimiter after it, allowing `:` in the `name-part` of external types. I will update the validation algorithm.

EDIT: created #511 to track this.

**Comment by zolkis** [comment]
> The only other bit I would flag, and I fully admit I'm not an expert here and so you'll want to find someone better versed in the Web Platform idioms, is where/how an error should be signaled for an unsupported recordType. Right now, if I understand correctly, I can create a 'valid' NDEFRecord via NDEFRecordInit, but whose recordType is invalid/unparsable. That error is not actually signaled back to the consumer until they try NDEFWriter.push, which invokes the create NDEF message algorithm, which invokes create NDEF record, which invokes validate external type.
>
> Because you expose the constructor for NDEFRecord, should you be parsing/validating the type then? The URL class, for better or worse, has setters which do the parsing/validation. I'm not sure why the NDEFRecord type is read-only (maybe that's the current best practice), but it seems that you at least want to validate during that constructor as well, to ensure a client only encounters well-formed NDEFRecords

Created #510 to track this.

**Comment by zolkis** [comment]
For the context, [this comment](https://github.com/w3c/web-nfc/issues/511#issuecomment-570620933) also belongs here.

To my understanding we can close this issue now. 
@beaufortfrancois if you agree, please do that, otherwise keep it open.
